server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java

```diff
@@ -52,6 +52,23 @@ public void initialize(ServerContext context) {
     this.context = context;
     zooCache = new ZooCache(context.getZooReaderWriter(), null);
     ZKUserPath = Constants.ZROOT + "/" + context.getInstanceID() + "/users";
+    checkOutdatedHashes();
+  }
+
+  private void checkOutdatedHashes() {
+    try {
+      listUsers().forEach(user -> {
+        String zpath = ZKUserPath + "/" + user;
+        byte[] zkData = zooCache.get(zpath);
+        if (ZKSecurityTool.isOutdatedPass(zkData)) {
+          log.warn("Found user(s) with outdated password hash. These will be re-hashed"
+              + " on successful authentication.");
```

Comment:
since we aren't including any details about which user(s) are impacted, I'd rather we not get a WARN for each user. could we move this to after we finish iterating with a summary of how many users?

If an operator needed to move towards eliminating these warn messages, how would they get the list of users that need to authenticate to the system?

Reply:
I'm never sure exactly how much detail is too much when it comes to logging security features. I can add an affected user count.
I'm not sure the operator has to or should do anything, the update is automatic. But with their current tools they could replicate what the code does and could go through the users zNode checking hash lengths. Alternately I could list the affected user principals in the warning but that feels unsafe. Do you have something in mind?

Reply:
The list of users in the server logs is okay. However, I just realized... this method is only ever called when we're starting from an empty user database (on initialize... or re-initialize), so there shouldn't be any matching the old hashes anyway at the point this code is run. Or did I miss something?

I do think maybe it would be worth the master server checking to see if any users are using the old hashing, and listing them in the server logs on startup, as an upgrade warning. In future, we can turn that into an upgrade blocker (so we can eventually drop the code that supports reading the old hashes).

Reply:
Added the check to master startup through SecurityOperation.

Reply:
@BukrosSzabolcs Feel free to mark the conversation as "Resolved" if you've taken steps to address the comments. It cleans up the interface and makes it easier for subsequent reviews. :smiley_cat: